website/main/content/blog/2025-04-28-proving-blocks-and-segments.md

```diff
@@ -0,0 +1,119 @@
+enables us to move faster in the design focusing on the parts that are still under-defined. This
+will also help us move from code to spec once we are comfortable with the design, as we just need to
+follow the reference in this spec outline and translate the implementation code into a spec _"(hey,
+we may even be able to delegate this task to an AI in the not-so-distant future)"_.
+
+## Proving blocks and segments belong to the history of the system.
+
+One of those parts of the protocol that were currently under-defined and that were blocking the
+implementation, were the proofs to verify that blocks and segments of a shard belong to the global
+history of the system. After some feedback iterations, this has been my main focus for the last
+week, and I am proud to share that we may have a first candidate design for these.
+
+### Block proofs
+
+Let's start with blocks. If you recall from previous updates, shards from the lower levels of the
+hierarchy are periodically committing the new blocks they are creating to their parents. This makes
+their history available to the rest of the system before a new segment of shard is created and made
+available in the global history maintained by the beacon chain (which make take some time to be
+available). These commitments would allow nodes to proof (and verify) that a block belongs to the
+history of the shard, and is consistent with the global history of the rest of the shards and the
+global history of the system.
+
+The following diagram shows a high-level of how I am thinking about the proofs for blocks. The idea
+is to use a
+[Merkle Mountain Range (MMR)](https://docs.grin.mw/wiki/chain-state/merkle-mountain-range/) to store
+the history of the blocks in the shard, and use references to blocks in the beacon chain to attach
+them to the global history in the beacon chain. This allows for the generation of incremental proofs
+of inclusion of blocks in the history of the shard, and their consistency with the global history.
+Child shards are periodically sending their block headers to their parent. Parent shards keep track
+of child shard new blocks, and they build their own view of the history of the shard. New blocks in
+shard need to always point to recent blocks in the beacon chain that can be consider final.
+
+Thus, when a network participant wants to prove that an event has happened, and this proof relies on
+a block in a child shard being valid, they can generate a proof of inclusion of a shard block by
+requesting a MMR proof for the block from the shard's history kept by the parent, and verifying that
+is consistent with the history of the parent shard, and that they point to valid blocks in the
+beacon chain (see the bottom-right part of the diagram for the high-level steps of the proof
+generation/verification).
+
+<p align="center">
+<img alt="High-level diagram of block proofs" src="mmr-block-proofs.png"></img>
```

Comment:
BTW I'd use a less fancy font on these diagrams going forward to make them easier to read

Reply:
Noted. This is excalidraw's default font. I think they have more "traditional" fonts. Will try with one of those the next time.